Configure enabled protocols per endpoint and add HTTP/2 w/ prior knowledge support (#2067)

cesarbs opened the pull request:
#2067

cesarbs commented:
Still kind of a WIP - need to discuss the design and add more tests.

davidfowl commented on src/Kestrel.Core/EndPointProtocols.cs:
```diff
@@ -0,0 +1,14 @@
+namespace Microsoft.AspNetCore.Server.Kestrel.Core
+{
+    [Flags]
+    public enum EndPointProtocols
```
HttpProtocols

Tratcher replied:
Or just Protocols? In the specs they are ApplicationProtocols.

Tratcher replied:
No combined value?

Tratcher replied:
Flags enums usually have a default `None = 0` value.

anurse replied:
Yeah, I'd add `None` and `Http1AndHttp2` (don't call it `Both`, as that means we're screwed if we need to add more later ;)). `None` is a little weird because if you specify it, we should throw out an error, but it can be useful for initialization, comparison, null values, etc.

cesarbs replied:
I'm aware flags typically have a `None`, but it doesn't make sense here. Why add it just to check for it and throw? I don't see it being useful for the cases you pointed out (when would that happen?).

`Http1AndHttp2` is ugly but I'm willing to put it there. I don't see what's wrong with `|` though.

Tratcher replied:
The default value for enum fields is 0, None provides a name for that value. That way it's more obvious when you're working with an uninitialized field and you don't have to check for 0 in code.

anurse replied:
`Http1AndHttp2` is definitely ugly. I'm not entirely opposed to encouraging `|`, but most of our (the greater 'our', the one that includes .NET ;)) flags enums have common combinations as specific values and `Both` is limiting.

Regarding `None`, @Tratcher basically has the reasoning. Since you can cast any int to an enum and `0` is a common enough value for uninitialized enums/ints, it's very useful for debugging purposes to have a name for it. If you don't want an explicit error case, that's OK. I'm sure some kind of error would occur if neither flag was set. Specifying `None` can have the same error case as specifying `(HttpProtocol)42`, but having the name is a useful debugging error.

anurse commented on src/Kestrel.Core/Internal/HttpConnection.cs:
```diff
@@ -143,15 +144,30 @@ public Task StartRequestProcessing<TContext>(IHttpApplication<TContext> applicat
+
+                if (!hasTls && http1Enabled && http2Enabled)
+                {
+                    throw new InvalidOperationException("HTTP/1.x and HTTP/2 cannot both be served on the same cleartext endpoint.");
```
Suggest rephrase: "Using both HTTP/1.x and HTTP/2 on the same endpoint requires the use of TLS"

anurse commented on src/Kestrel.Core/Internal/HttpConnection.cs:
```diff
@@ -143,15 +144,30 @@ public Task StartRequestProcessing<TContext>(IHttpApplication<TContext> applicat
+                    throw new InvalidOperationException("HTTP/1.x and HTTP/2 cannot both be served on the same cleartext endpoint.");
+                }
+
+                var useHttp2 = http2Enabled && (!hasTls || applicationProtocol == "h2");
```
Should we use byte arrays for these values? As discussed with @Tratcher yesterday, ALPN just uses byte strings for protocols rather than specifying strings and encodings and such.

Tratcher replied:
I'm not immediately concerned with this as long as it doesn't leak into public APIs. We'll have to refactor when we get the real ALPN support anyways.

anurse replied:
`ITlsApplicationProtocolFeature` is (and will be released someday as) a public API. It's a generic ALPN feature for TLS connections.

Tratcher replied:
There's a new type coming that we can use for the datatype here so it doesn't have to be raw bytes.
https://github.com/dotnet/corefx/issues/23177 or we can go down one layer to `ReadOnlyMemory<byte>`.

anurse replied:
I think we'd need to go to `ReadOnlyMemory<byte>`. We'd want this interface to be `netstandard` even though the implementation will be `netcoreapp2.x` until the SslStream features reach `netstandard`. So unless the exchange type is moved to a different `netstandard`-compatible assembly, I think we'll have to bypass it.

anurse commented on src/Kestrel.Core/Internal/HttpConnection.cs:
```diff
@@ -143,15 +144,30 @@ public Task StartRequestProcessing<TContext>(IHttpApplication<TContext> applicat
                 }
+                else
+                {
+                    throw new InvalidOperationException("Unable to select a protocol for the endpoint.");
```
I generally dislike "Unable to do thing" messages that don't say why. If a few more `if` clauses are needed to figure out a more specific error, I say we do that work so consumers don't have to.

cesarbs commented:
I've addressed most of your feedback. I'm now working on setting up ALPN on `TlsStream` based on the value of `ListenOptions.Protocols`.

cesarbs commented:
Latest commit uses `ListenOptions.Protocols` to set up ALPN on `TlsStream`.

Tratcher commented on test/Kestrel.FunctionalTests/Http2Tests.cs:
```diff
@@ -0,0 +1,42 @@
+    public class Http2Tests
+    {
+        [Fact]
+        public async Task ServesHttp2WithPriorKnowledge()
```
No ALPN test? or do you not have an ALPN client?

cesarbs replied:
Currently working on more tests. Not sure I'll be able to final an ALPN client.

Tratcher commented on src/Kestrel.Tls/TlsConnectionAdapter.cs:
```diff
@@ -47,6 +47,17 @@ public TlsConnectionAdapter(TlsConnectionAdapterOptions options, ILoggerFactory
             _logger = loggerFactory?.CreateLogger(nameof(TlsConnectionAdapter));
+
+            // Order is important. If HTTP/2 is enabled, we prefer it over HTTP/1.1. So add it first.
+            if ((options.Protocols & HttpProtocols.Http2) == HttpProtocols.Http2)
```
You could #ifdef this section in preparation for the new core API. Then you'd be able to test the net46 behavior.

cesarbs replied:
Since I can't properly test with ALPN now, I'll do that in a future change.

cesarbs commented:
I've renamed `Http2Tests` to `HttpProtocolSelectionTests`. For now there are only cleartext tests there since I couldn't make `HttpClient` use HTTP/2 over TLS on neither Windows nor Linux. We can add TLS tests once we have ALPN on `SslStream`.

I've added `ListenOptionsTests` with a test to check that endpoints are HTTP/1.x-only by default.

cesarbs commented on src/Kestrel.Core/Internal/HttpConnection.cs:
```diff
@@ -170,17 +170,15 @@ public Task StartRequestProcessing<TContext>(IHttpApplication<TContext> applicat
             finally
             {
+                // Complete the output in case an exception was thrown before request processing could start.
+                _context.Transport.Output.Complete();
```
@davidfowl I'd like your input here.

cesarbs commented:
Ping.

Tratcher commented on src/Kestrel.Core/Internal/HttpConnectionBuilderExtensions.cs:
```diff
@@ -11,12 +11,22 @@ public static class HttpConnectionBuilderExtensions
         public static IConnectionBuilder UseHttpServer<TContext>(this IConnectionBuilder builder, ServiceContext serviceContext, IHttpApplication<TContext> application)
         {
-            return builder.UseHttpServer(Array.Empty<IConnectionAdapter>(), serviceContext, application);
+            return builder.UseHttpServer(serviceContext, application, protocols: HttpProtocols.Http1);
```
why the named parameters here?

cesarbs replied:
Talked to @davidfowl. I'm removing those extension methods and leaving only the new ones (with the new `HttpProtocols` param).

davidfowl commented on src/Kestrel.Core/Internal/HttpConnection.cs:
```diff
@@ -290,16 +297,50 @@ private void DisposeAdaptedConnections()
+            if (error != null)
+            {
+                Log.LogError(0, error);
+                Abort(ex: null);
```
SIDE EFFECTS! Do this outside.
